internal/stream/event_stream.go

```diff
@@ -0,0 +1,94 @@
+package stream
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/openshift/assisted-service/models"
+	"github.com/sirupsen/logrus"
+)
+
+type EventStream struct {
+	writer EventStreamWriter
+	log    logrus.FieldLogger
+}
+
+const (
+	TypeInfraEnv = "InfraEnv"
+	TypeHost     = "HostState"
+	TypeCluster  = "ClusterState"
+	TypeEvent    = "Event"
+)
+
+type Resource interface{}
+
+func NewEventStream(writer EventStreamWriter, logger logrus.FieldLogger) *EventStream {
+	return &EventStream{
+		writer: writer,
+		log:    logger,
+	}
+}
+
+func (s *EventStream) Notify(ctx context.Context, resource Resource) error {
+	if s.writer == nil || resource == nil {
+		return nil
+	}
+	key, eventType, logFields, err := getNotificationDetails(resource)
+	if err != nil {
+		s.log.WithError(err).WithFields(logrus.Fields{
+			"resource": resource,
+		}).Warn("failed to detect resource type or resource nil")
+		return err
+	}
+	err = s.writer.Write(ctx, eventType, []byte(key), resource)
+	if err != nil {
+		s.log.WithError(err).WithFields(logFields).Warn("failed to stream event for infraenv")
+	}
+	return err
+}
+
+func (s *EventStream) Close() {
+	if s.writer != nil {
+		s.writer.Close()
+	}
+}
+
+// Returns key, event type and fields to be logged
+func getNotificationDetails(resource Resource) (string, string, logrus.Fields, error) {
```

Comment:
This looks like a really good place to actually use interfaces.

We should define an interface with this method on it that can be implemented by `common.<Thing>`.

Reply:
In the end this approach helped us extract most of the logic from kafka writer, leaving it as a basic wrapper for configuration and JSON Encoding. I would still leave it as separate, as it can be replaced for any other encoding we might want to use or separate storage for notifications